packages/lib/EventDispatcher.ts

```diff
@@ -0,0 +1,71 @@
+type Listener<Value> = (data: Value)=> void;
+
+/**
+ * Sends notifications/allows listening for events.
+ *
+ * @param EventKeyType Possible identifiers for events.
+ * @param EventType Type of data associated with events.
+ */
+export default class EventDispatcher<EventKeyType extends string|symbol|number, EventType> {
+	// Partial marks all fields as optional. To initialize with an empty object, this is required.
+	// See https://stackoverflow.com/a/64526384
+	private listeners: Partial<Record<EventKeyType, Array<Listener<EventType>>>>;
+	public constructor() {
+		this.listeners = {};
+	}
+
+	/**
+	 * Sends a notification to all listeners subscribed to the given [notificationKind].
+	 * @param eventType Type of the notification.
+	 * @param value Data associated with the notification.
+	 */
+	dispatch(eventName: EventKeyType, event: EventType = null) {
+		if (!this.listeners[eventName]) return;
+
+		const ls = this.listeners[eventName];
+		for (let i = 0; i < ls.length; i++) {
+			ls[i](event);
+		}
+	}
+
+	/**
+	 * Listen for a specific type of event.
+	 *
+	 * @param eventType Type of event that causes `callback` to be called.
+	 * @param callback Callback called when an event of `eventType` is fired.
+	 * @returns An object with a `remove` function. Calling `remove` unregisters the listener.
+	 */
```

Comment:
You don't need to add this JavaDoc stuff. TypeScript already allows specifying the type in a better way, and your code in general should be self-documenting. Only add a comment when it's difficult to express the intention in code.

Reply:
VSCode has good support for JSDoc (what I've been trying to use) — hovering over a method name shows JSDoc comment content (but not `//` comment content).

Of course, this is IDE-specific and a stylistic preference. I can remove the JSDoc comments.

Reply:
Changed!